File: app/src/main/java/Day25.java
import shared.SimpleStack;

import java.util.Arrays;
import java.util.NoSuchElementException;

public class Day25 {

    public static <E> SimpleStack<E> getSimpleStack() {

        return new SimpleStack<E>() {
            //An ArrayList already handles resizing and supports generics,
            //but we'll do a T array instead to address more of the challenge
            int top = -1;  //current location of top of stack
            @SuppressWarnings("unchecked")
            E[] stack = (E[]) new Object[BLOCK];  //initialize the array to the first block size

            @Override
            public void push(E item) {
                if (top + 1 >= MAX_BLOCK) {
                    throw new IllegalStateException();
                }

                if (top + 1 >= stack.length) {
                    // resize the array within bounds
                    int newLen = Math.min(stack.length + BLOCK, MAX_BLOCK);
                    stack = Arrays.copyOf(stack, newLen);
                }

                stack[++top] = item;
            }

            @Override
            public E pop() {
                if (top < 0) {
                    throw new NoSuchElementException();
                }

                E item = stack[top];
                //remove the item reference to allow GC, prevent memory leak
                stack[top] = null;
                top -= 1;
                return item;
            }

            @Override
            public E peek() {
                if (top < 0) {
                    throw new NoSuchElementException();
                }

                return stack[top];
            }

            @Override
            public boolean isEmpty() {
                return top < 0;
            }
        };
    }
}

File: app/src/main/kotlin/Day25.kt
import shared.SimpleStack
import shared.SimpleStack.BLOCK
import shared.SimpleStack.MAX_BLOCK

// JVM-visible wrapper
fun getSimpleStack(): SimpleStack<Any> = getSimpleStackInline()

inline fun <reified E> getSimpleStackInline(): SimpleStack<E> {
    return object : SimpleStack<E> {
        var top = -1
        var stack: Array<E?> = Array(BLOCK) { null }

        override fun push(item: E) {
            if (top + 1 >= MAX_BLOCK) throw IllegalStateException()

            if (top + 1 >= stack.size) {
                val newLen = kotlin.math.min(stack.size + BLOCK, MAX_BLOCK)
                stack = stack.copyOf(newLen)
            }

            stack[++top] = item
        }

        override fun pop(): E {
            if (top < 0) throw NoSuchElementException()
            val item = stack[top] ?: throw NoSuchElementException()
            stack[top] = null
            top--
            return item
        }

        override fun peek(): E {
            if (top < 0) throw NoSuchElementException()
            return stack[top]!!
        }

        override fun isEmpty(): Boolean = top < 0
    }
}